clients/src/main/java/org/apache/kafka/common/security/authenticator/DefaultKafkaPrincipalBuilder.java

```diff
@@ -167,4 +171,24 @@ public void close() {
             oldPrincipalBuilder.close();
     }
 
+    @Override
+    public ByteBuffer serialize(KafkaPrincipal principal) {
+        DefaultPrincipalData data = new DefaultPrincipalData()
+                                        .setType(principal.getPrincipalType())
+                                        .setName(principal.getName())
+                                        .setTokenAuthenticated(principal.tokenAuthenticated());
+        Struct dataStruct = data.toStruct(DefaultPrincipalData.HIGHEST_SUPPORTED_VERSION);
+        ByteBuffer buffer = ByteBuffer.allocate(dataStruct.sizeOf());
+        dataStruct.writeTo(buffer);
+        buffer.flip();
+        return buffer;
+    }
+
+    @Override
+    public KafkaPrincipal deserialize(ByteBuffer bytes) {
+        DefaultPrincipalData data = new DefaultPrincipalData(
+            DefaultPrincipalData.SCHEMAS[DefaultPrincipalData.SCHEMAS.length - 1].read(bytes),
+            DefaultPrincipalData.HIGHEST_SUPPORTED_VERSION);
```

Comment:
Rather than assuming highest supported version, we should include the version in the serialized data. The simple thing would be to write the version first, then write the payload.